WordPress/gutenberg
# Site Editor: Prevent unintended actions on the classic theme

Related to #54066, #52150
Related core ticket: https://core.trac.wordpress.org/ticket/58827
Related core PR: https://github.com/WordPress/wordpress-develop/pull/5201#issuecomment-1726991594

## What?
This PR prevents the following three unintended actions in case the Classic theme is allowed to access the Patterns page of the Site Editor.

- Can create a template part
- Can back to Design page
- Can access All template parts page
- The back button link on the All Template Parts page is to the dashboard even if the Site Editor's Patterns page is accessed directly (in this scenario, it is expected to return to the Patterns page)

| Before | After |
|--------|--------|
| ![before](https://github.com/WordPress/gutenberg/assets/54422211/e7091c2f-3c79-46bd-aafb-1c17538b58b4) | ![after](https://github.com/WordPress/gutenberg/assets/54422211/9955e553-e03a-460e-8445-79a12a230bdf) |

## Why?

As a first step toward making the Patterns page (`site-editor.php?path=/patterns`) accessible to all classic themes in the future, we are considering the following in WP6.4:

1. Add an `Appearance -> Patterns` link for all classic themes that point to `edit.php?post_type=wp_block`.
2. Allow `site-editor.php?path=/patterns` access for classic themes internally.

This means that users of the classic theme will not see the link to the Pattern page (`site-editor.php?path=/patterns`) anywhere, but will be able to access it by typing its URL directly into their browser.

Therefore, actions that are not expected in that case need to be prevented in advance.

## Testing Instructions

To temporarily allow access to the site editor from the classic theme, comment out [here](https://github.com/WordPress/wordpress-develop/blob/54a177910a80083930ee1284f648c1db5bf2e5c0/src/wp-admin/site-editor.php#L22-L32) in the WordPress core

1. Access `http://localhost:8889/wp-admin`
2. Activate Twenty Twenty One
3. Enter `http://localhost:8889/wp-admin/site-editor.php?path=%2Fpatterns` in the URL bar of your browser to access the Patterns page
4. Make sure the "Manage all template parts" menu is not displayed
5. Confirm that new template parts cannot be created
6. Check these steps also for the classic theme (EmptyHybrid) with block-template-parts support

### Additional Test

If a classic theme that supports block template parts goes directly to the Patterns page in the Site Editor, the back button link on the All Template Parts page should always go to the Patterns page instead of the Dashboard.

To control this, I introduce a URL parameter called `didAccessPatternsPage`.

Based on this, please test the following:

- Block Theme: It should work as before. No matter what you do, the URL should always have no `didAccessPatternsPage` parameter.
- Classic Theme _with_ template parts support:
  - Activate Emptyhybrid theme.
  - Access `http://localhost:8889/wp-admin/site-editor.php?path=%2Fpatterns` directly.
  - No matter what operation is performed, the All Template Parts page should return to the Patterns page.
  - The `didAccessPatternsPage` URL parameter is retained even if the browser is reloaded.
  - You should always return to the Patterns page, even after reloading the browser or reviving a closed tab with `Cmd+Shift+T`.
- Classic Theme _without_ template parts support:
  - Activate Twenty Twenty-One.
  - Access `http://localhost:8889/wp-admin/site-editor.php?path=%2Fpatterns` directly.
  - The URL should not have a `didAccessPatternsPage` parameter, no matter what operation is performed, not to mention that the template parts are not accessible.
